### PR TITLE
Hotreload extension

### PR DESCRIPTION
**Pull request by juanolon**
When this extension is enable, a websocket endpoint is created at `/ws`. The websocket will publish wjhich page changed, and the client will redirect the page to the changed page (#52 )

### CODE REVIEW

**Comment by juanolon** [comment]
What i wasn't able to do, was to get the current server host. right now, the websocket endpoint is hardcoded in javascript to localhost:3000

File: extensions/hotreload/hotreload.go
Comment (emad-elsaid) [inline extensions/hotreload/hotreload.go:92]: evt.data at this point is a string, so `.url` is undefined. so I think you'll need to move line 93 before 92 to parse first then use `data.url`

The other point is `window.location.href` returns the  whole URL `http://localhost:3000/pagename` not `/pagename` and the `url` you passed in the event is just `/pagename` so it'll never match

but also what if it matches? shouldn't we reload the page to show the changes? 

So I think this section should be without this if statement, and directly parse and set the href. it should navigate if the page is different and reload if not. and I think we can use `.path` instead of `href`

File: extensions/hotreload/hotreload.go
Comment (emad-elsaid) [inline extensions/hotreload/hotreload.go:9]: there is no need to import "embed" here as we don't use it in the extension

File: extensions/hotreload/hotreload.go
Comment (emad-elsaid) [inline extensions/hotreload/hotreload.go:22]: `READONLY` will be always false, as when `init()` is called the flags are not parsed yet. so I think you'll need to check `READONLY` variable inside each `handleWebSocket` and `clientWidget` to decide if they should return early.

File: extensions/hotreload/hotreload.go
Comment (emad-elsaid) [inline extensions/hotreload/hotreload.go:52]: :eyes: I'm not sure if this is required ? Do you want maybe `xlog.NoContent` ?

File: extensions/hotreload/hotreload.go
Comment (emad-elsaid) [inline extensions/hotreload/hotreload.go:89]: I think `window.location.host` returns the `localhost:3000` part.

File: markdown_fs.go
Comment (emad-elsaid) [inline markdown_fs.go:72]: I noticed the indentation is not consistent here. I see in my editor that it uses spaces and other lines uses tabs. maybe editor config needs adjusting. I guess `go fmt` needs to run on this file?

**Comment by emad-elsaid** [comment]
if you'd like me to do these changes let me know

File: extensions/hotreload/hotreload.go
Comment (juanolon) [reply extensions/hotreload/hotreload.go:92]: oh yes. that's true. and yes, better without the if statement. I'm already using it as it is, and it reloads every time, as expected not expected

File: extensions/hotreload/hotreload.go
Comment (juanolon) [reply extensions/hotreload/hotreload.go:52]: `xlog.NoContent` injects a http header to the response, then not being able to convert the request to a websocket. so I just made a new real nop

Comment (emad-elsaid) [reply extensions/hotreload/hotreload.go:48]: I see, Can we add this to `server.go` and export it as `Noop` so it can be used in other cases?

File: extensions/hotreload/hotreload.go
Comment (emad-elsaid) [inline extensions/hotreload/hotreload.go:99]: I guess we don't need this comment and the other one in L84 anymore?

**Review by emad-elsaid — APPROVED** [review]

**Comment by emad-elsaid** [comment]
I'll merge it and release a new version after we merge the mathJax PR

**Comment by juanolon** [comment]
You're welcome! that sounds nice!

**Pull request merged**
